JIT: Finally chain merging

AndyAyersMS opened the pull request:
Add an optimization that performs a specialized tail merge for chains
of callfinallys. These can arise from try-finallys where there are
multiple exit points from the try that have the same continuation.

When the jit generates a callfinally, it does so without considering
whether some previously generated callfinally might invoke the same
handler and then continue on to the same continuation.

This optimization looks for callfinallys that invoke the same handler
and have the same continuation and effectively merges them into one
canonical callfinally.

This optimization is done bottom-up (from outermost handler invocation
to innermost) to allow the merging done at an outer levels to enable
merging at inner levels.

This optimization saves code size and also results in more compact EH
reporting, since the callfinallies are reported to the runtime as
duplicate EH regions.

When run upstream of finally cloning, finally chain merging also allows
more of the paths exiting the try to exit via the clone.

AndyAyersMS commented:
This can stand on its own, so prepping it now for some early testing and feedback. I plan to hold off on finishing this until after finally cloning is merged.

Code size impact as a standalone change:
```
Total bytes of diff: -9857 (-0.07 % of base)
    diff is an improvement.
Total byte diff includes 0 bytes from reconciling methods
        Base had    0 unique methods,        0 unique bytes
        Diff had    0 unique methods,        0 unique bytes
Top file improvements by size (bytes):
       -1600 : System.Private.CoreLib.dasm (-0.04 % of base)
       -1353 : Microsoft.CodeAnalysis.CSharp.dasm (-0.06 % of base)
       -1233 : Microsoft.CodeAnalysis.VisualBasic.dasm (-0.05 % of base)
        -669 : System.Linq.Parallel.dasm (-0.11 % of base)
        -659 : System.Threading.Tasks.Dataflow.dasm (-0.44 % of base)
46 total files with size differences.
Top method improvements by size (bytes):
        -318 : Microsoft.CodeAnalysis.CSharp.dasm - Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.LanguageParser:ParseMemberDeclarationOrStatement(char,ref):ref:this
        -177 : Microsoft.CodeAnalysis.VisualBasic.dasm - Microsoft.CodeAnalysis.VisualBasic.Binder:BindMemberAccess(ref,ref,ref,bool,bool,ref):ref:this
        -171 : Microsoft.CodeAnalysis.VisualBasic.dasm - Microsoft.CodeAnalysis.VisualBasic.Binder:ReportOverloadResolutionFailureForASingleCandidate(ref,ref,int,byref,struct,struct,bool,bool,bool,bool,ref,ref,bool,ref,ref):this
        -152 : System.Security.Cryptography.X509Certificates.dasm - Internal.Cryptography.Pal.FindPal:FindFromCollection(ref,int,ref,bool):ref
        -144 : Microsoft.CodeAnalysis.dasm - Microsoft.CodeAnalysis.CommonCompiler:RunCore(ref,ref,struct):int:this
507 total methods with size differences.
```

Sample code diff:
```
;;; Microsoft.Win32.RegistryKey:Dispose(bool):this
;;; Before
...
G_M13601_IG05:
       mov      rcx, rsp
       call     G_M13601_IG18
G_M13601_IG06:
       jmp      G_M13601_IG13
G_M13601_IG07:
       mov      rcx, rsp
       call     G_M13601_IG18
G_M13601_IG08:
       jmp      G_M13601_IG13
...
2 EH table entries, 1 duplicate clauses, 2 cloned finallys, 5 total EH entries reported to VM
EH#0: try [G_M13601_IG03..G_M13601_IG04) handled by [G_M13601_IG15..G_M13601_IG18) (class: 200022E)
EH#1: try [G_M13601_IG03..G_M13601_IG05) handled by [G_M13601_IG18..END) (finally)
EH#2: try [G_M13601_IG15..G_M13601_IG18) handled by [G_M13601_IG18..END) (finally) duplicated
EH#3: try [G_M13601_IG05..G_M13601_IG05) handled by [G_M13601_IG05..G_M13601_IG07) (finally) cloned finally
EH#4: try [G_M13601_IG07..G_M13601_IG07) handled by [G_M13601_IG07..G_M13601_IG09) (finally) cloned finally

;;; After
... (branches retargeted upstream)
G_M13601_IG05:
       mov      rcx, rsp
       call     G_M13601_IG16
G_M13601_IG06:
       jmp      G_M13601_IG11
...
2 EH table entries, 1 duplicate clauses, 1 cloned finallys, 4 total EH entries reported to VM
EH#0: try [G_M13601_IG03..G_M13601_IG04) handled by [G_M13601_IG13..G_M13601_IG16) (class: 200022E)
EH#1: try [G_M13601_IG03..G_M13601_IG05) handled by [G_M13601_IG16..END) (finally)
EH#2: try [G_M13601_IG13..G_M13601_IG16) handled by [G_M13601_IG16..END) (finally) duplicated
EH#3: try [G_M13601_IG05..G_M13601_IG05) handled by [G_M13601_IG05..G_M13601_IG07) (finally) cloned finally

```

AndyAyersMS commented:
Suspect this [arm failure] (https://ci.dot.net/job/dotnet_coreclr/job/master/job/arm_emulator_cross_release_ubuntu_prtest/1478/) is spurious, will rerun.

@dotnet-bot retest Linux ARM Emulator Cross Release Build

AndyAyersMS commented:
Same two tests failed again. Looked at one and it has some try/finallys, so will investigate.

AndyAyersMS commented:
Think the failures may be related to improper maintenance of the arm-only finally target bit. Have modified the two tests that were failing to always build with optimization enabled so that the jit tries the same codegen in debug as it does in release; let's see if that triggers any assertions/failures on the debug run.

AndyAyersMS commented:
@leemgs any further advice on how to set up the arm emulator environment locally? I hate to use the lab for experimentation. 

I have Ubuntu 14.04 and have tried to follow the notes in the cross build script faithfully, but things aren't working. I can get the cross build to go through but when it executes runtest.sh, it complains (among other things) that cp doesn't have a -n option and that there is no perl.

AndyAyersMS commented:
@sjsinju / @wateret if you can help that would also be appreciated....

BruceForstall commented on src/jit/flowgraph.cpp:
```diff
@@ -22471,3 +22471,240 @@ void Compiler::fgLclFldAssign(unsigned lclNum)
+        // try/fault but presumably those are rare.
+        if (HBtab->HasFinallyHandler())
+        {
+            finallyCount++;
```
Looks like you don't use the count, so `finallyCount` could be `bool hasFinallies`, and once you find one you could break out of the loop.

AndyAyersMS replied:
Will fix

BruceForstall commented on src/jit/flowgraph.cpp:
```diff
@@ -22471,3 +22471,240 @@ void Compiler::fgLclFldAssign(unsigned lclNum)
+    if (!canMerge)
+    {
+        JITDUMP("Method had try-finallys, but did not have any mergeable finally chains.\n");
+        return;
```
Might be better to have the remaining DEBUG code in an `else` clause to avoid requiring this easy-to-miss-seeing `return`.

AndyAyersMS replied:
Yep, restructured this bit.

BruceForstall commented on src/jit/flowgraph.cpp:
```diff
@@ -22471,3 +22471,240 @@ void Compiler::fgLclFldAssign(unsigned lclNum)
+
+        // Build a map from each continuation to the "canonical"
+        // callfinally for that continuation.
+        BlockToBlockMap   continuationMap(getAllocator());
```
It might be better to declare continuationMap outside the loop, once, and then use `continuationMap.RemoveAll()` inside the loop, for better memory/allocation usage.

AndyAyersMS replied:
Sure.

BruceForstall commented on src/jit/flowgraph.cpp:
```diff
@@ -22471,3 +22471,240 @@ void Compiler::fgLclFldAssign(unsigned lclNum)
+        // and make sure they all jump to the appropriate canonical
+        // callfinally.
+        BasicBlock* const endTryBlock = HBtab->ebdTryLast->bbNext;
+        for (BasicBlock* currentBlock = HBtab->ebdTryBeg; currentBlock != endTryBlock;
```
I don't think this first loop (over the `try` region) is necessary: the "call finally" region will be identical to this in the `!FEATURE_EH_CALLFINALLY_THUNKS` case (all but x64), and in the `FEATURE_EH_CALLFINALLY_THUNKS` case, to call finally range is (I believe) always a superset.

BruceForstall replied:
(oh, I guess ARM64 uses the `FEATURE_EH_CALLFINALLY_THUNKS` code, too)

AndyAyersMS replied:
I wasn't sure about the thunks case, but you're right and the walk over the try is not needed. Will remove.

BruceForstall commented on src/jit/flowgraph.cpp:
```diff
@@ -22471,3 +22471,240 @@ void Compiler::fgLclFldAssign(unsigned lclNum)
+        // Possible paranoia assert here -- no flow successor of
+        // this block should be a callfinally for this try.
+        return false;
+    }
```
This condition holds for AMD64 (and probably ARM64) but not for ARM32. For ARM32, there is no branch to a CALLFINALLY/ALWAYS pair "thunk" (that is placed outside the 'try' region). Instead, the CALLFINALLY/ALWAYS pair is in the location of the actual 'try' region exit, within the 'try' region itself. Because the code sequence we generate on ARM32 never returns, this is ok.

BruceForstall replied:
btw, you can decode that from the `!FEATURE_EH_CALLFINALLY_THUNKS` case in `impImportLeave()` handling the insertion of the BBJ_CALLFINALLY, around line 8464 "We are jumping out of a finally-protected try".

BruceForstall replied:
So, on arm32, if you have:
```
try1 {
   try2 {
      callfinally F2 / always L1 // A
      ...
      callfinally F2 / always L2 // B
      ...
   }
   L1: callfinally F1 / always L3 // C
   L2: callfinally F1 / always L3 // D
}
L3:

finally2: ... // F2
finally1: ... // F1
```

So we want to merge L1 & L2. Say L1 is canonical. When processing the 'always' of 'B' you would change it point to L1. Interestingly, that seems like it should work, although I don't think that's what you were thinking the structure of the callfinally blocks was when you wrote the code. I wonder if there's some more complex case I'm missing.

On amd64 you would have seen:
```
try1 {
   try2 {
      always A1
      ...
      always A2
      ...
   }
   A1: callfinally F2 / always L1 // A
   A2: callfinally F2 / always L2 // B
}
L1: callfinally F1 / always L3 // C
L2: callfinally F1 / always L3 // D
L3:

finally2: ... // F2
finally1: ... // F1
```

AndyAyersMS replied:
I think you are saying that on ARM32 the address of the continuation is passed to the finally as the resume point, instead of the address of the always block. But the successor of the always block is the method the jit uses to identify the continuation, so changing the successor like this seems on the surface like it ought to work. And since we're only swapping one continuation for another the new successor should have the finally target bit set.

I still suspect something funny may be going on with the finally target bit, because chain merging will create more cases where a such a block has multiple predecessors. But no hard evidence yet.

AndyAyersMS replied:
Ah, perhaps you are saying that on in the non-thunk models the callfinally block may contain other code beside the jump the the finally. If so then we might need a more stringent check to safely merge. Currently we are implicitly assuming that the only thing that matters in a callfinally/leave/continuation block set is the identity of the handler and the identity of the continuation.

BruceForstall replied:
Ah, that could be a problem. In the thunk model, the BBJ_CALLFINALLY is in an empty block, but in the non-thunk model, that isn't necessarily so.

AndyAyersMS replied:
Agreed, that seems a likely suspect. Adding an assert now that will fire if the canonical callfinally is not an empty block.

BruceForstall commented:
BTW, you could try porting your changes to desktop and using our ngen-based ARM32 altjit asm diffs there to find cases to look at.

BruceForstall commented on src/jit/flowgraph.cpp:
```diff
@@ -22577,10 +22580,14 @@ void Compiler::fgMergeFinallyChains()
+                    // This callfinally should not contain any statements.
+                    // If it does, we can't safely cross-jump here. We
+                    // can fix this by splitting the block....
+                    assert(currentBlock->firstStmt() == nullptr);
```
you should use `currentBlock->isEmpty()`

AndyAyersMS replied:
Sure.

AndyAyersMS commented:
@BruceForstall not sure you will like this latest change but it should at least resolve whether callfinallys with other statements were responsible for the arm failures.

AndyAyersMS commented:
Or maybe not... I really need some kind of workable arm32 test setup.

BruceForstall commented:
I guess your change was speculative since you couldn't actually test on arm32?

AndyAyersMS commented:
Yes, speculative -- I'm going to look more closely at what happens on x86 and then try your suggestion looking at arm on desktop.

AndyAyersMS commented:
On x86, the continuation is (always?) a block with an `GT_END_LFIN` that jumps to the actual continuation. These blocks would also need to be merged in order to merge the callfinallys. So as-is, no merging happens on x86.

AndyAyersMS commented:
Not sure where to go with this just yet. Let me think about it some more.

The size improvement on x64 is small, but part of what appealed to me here was that the size savings basically would cancel out the size increase from finally cloning, so the net size impact of the finally optimization passes  (remove empty / merge chains / clone) is effectively zero.

BruceForstall commented:
I never remember what x86 does, but I think it accumulates chains of endfinally/endcatch tree node actions. It might be possible to merge these, but it would require different logic. I would avoid x86 (by using `FEATURE_EH_FUNCLETS`) to start with.

AndyAyersMS commented:
Rebased on top of the finally cloning changes, and added opt-outs for non-thunk and non-callfinally thunk EH models.

AndyAyersMS commented:
Current size win on x64 (smaller than before; not too surprising, given that empty finallys are now removed upstream):
```
Total bytes of diff: -2310 (-0.02 % of base)
    diff is an improvement.
Total byte diff includes 0 bytes from reconciling methods
        Base had    0 unique methods,        0 unique bytes
        Diff had    0 unique methods,        0 unique bytes
Top file regressions by size (bytes):
          14 : System.Net.Primitives.dasm (0.02 % of base)
Top file improvements by size (bytes):
        -530 : System.Private.CoreLib.dasm (-0.01 % of base)
        -313 : System.Linq.Parallel.dasm (-0.05 % of base)
        -276 : Microsoft.CodeAnalysis.VisualBasic.dasm (-0.01 % of base)
        -212 : Microsoft.CodeAnalysis.dasm (-0.03 % of base)
        -175 : System.Net.Security.dasm (-0.12 % of base)
24 total files with size differences.
Top method regessions by size (bytes):
          14 : System.Net.Primitives.dasm - System.Net.CookieContainer:BuildCookieCollectionFromDomainMatches(ref,bool,int,byref,ref,bool):this
           6 : Microsoft.CodeAnalysis.CSharp.dasm - Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.LanguageParser:ParseCastOrParenExpressionOrLambda(int):ref:this
           6 : System.Net.Security.dasm - System.Net.CertificateValidationPal:EnsureStoreOpened(bool):ref
           4 : Microsoft.CodeAnalysis.CSharp.dasm - Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.LanguageParser:ParseMemberName(byref,byref,byref,bool):this
Top method improvements by size (bytes):
         -80 : Microsoft.CodeAnalysis.dasm - Microsoft.CodeAnalysis.CommonCompiler:RunCore(ref,ref,struct):int:this
         -63 : Microsoft.CodeAnalysis.dasm - Microsoft.CodeAnalysis.Compilation:SerializeToPeStream(ref,ref,ref,ref,ref,bool,struct):bool:this
         -51 : System.Net.Requests.dasm - System.Net.FtpWebRequest:FinishRequestStage(int):int:this
         -49 : System.Security.Cryptography.X509Certificates.dasm - Internal.Cryptography.Pal.ChainPal:BuildChain(bool,ref,ref,ref,ref,int,int,struct,struct):ref
         -46 : Microsoft.CodeAnalysis.VisualBasic.dasm - Microsoft.CodeAnalysis.VisualBasic.VisualBasicCommandLineArguments:ResolveMetadataReferences(ref,ref,ref,ref):bool:this
156 total methods with size differences.
```

The earlier version of chain merging was run without finally cloning or empty finally removal and modified 507 methods. There are/were ~96 methods with empty finally so one might expect running this after empty finally we'd still see something 400 methods with diffs. However finally cloning is also able to straighten out some exit paths. I'll try running this with finally cloning disabled to see if the number pops back up to around 400.

AndyAyersMS commented:
Running this with empty finally enabled upstream and finally cloning disabled downstream ups the number of methods impacted by this to nearly 500. Presumably that means most of the 100 or so empty finally cases didn't have mergeable chains, and about 2/3 of the 500 or so cases with mergeable chains can were fixed by the simple one level merge that finally cloning does.

AndyAyersMS commented:
@BruceForstall "finally" brought this up to date with all the other EH opts.....

AndyAyersMS commented:
CI was down when I did the last push.

@dotnet-bot test this please

AndyAyersMS commented:
AutoResetEventTest failed [(log)](https://ci.dot.net/job/dotnet_coreclr/job/master/job/x64_release_windows_nt_pri1_prtest/2707/). Quite likely unrelated; looks like another instance of #8348.
```
13:24:01         Calling AutoResetEvent.WaitOne()... if the event is signaled it will not wait long enough
13:24:01         ERROR!!!-002: AutoResetEvent did not wait long enough... this implies that the parameter was not respected.
13:24:01         ERROR!!!-002:  WaitTime=83913310 (ticks)
13:24:01         ERROR!!!-002:  Execpted=50000000 (ticks)
13:24:01         ERROR!!!-002:  Acceptable Delta=9999999 (ticks)
```
@dotnet-bot retest Windows_NT x64 Release Priority 1 Build and Test

BruceForstall commented on src/jit/flowgraph.cpp:
```diff
@@ -22995,6 +22995,7 @@ void Compiler::fgRemoveEmptyTry()
                     block->bbJumpKind = BBJ_ALWAYS;
                     block->bbJumpDest = continuation;
+                    fgAddRefPred(continuation, block);
                 }
```
Is this related to this change?

It looks like preds haven't been computed by the time your try/finally optimization phases are run, so perhaps you should remove all this pred update stuff, and add (for clarity) the following at the top of each phase:
```
noway_assert(!fgComputePredsDone); 
```

AndyAyersMS replied:
Yes, this was a fix necessitated by this change.

From what I can tell all the ref count updates are accurate right now (with the exception of the finally entry block, which is a special case). It would seem odd to deliberately leave them inaccurate.

AndyAyersMS replied:
Looks like I need to rebase to get past a merge conflict, so I'll also add the assert.

BruceForstall commented on src/jit/flowgraph.cpp:
```diff
@@ -23873,6 +23874,263 @@ void Compiler::fgUpdateFinallyTargetFlags()
+    bool            canMerge = false;
+    bool            didMerge = false;
+    BlockToBlockMap continuationMap(getAllocator());
+
```
Is it worth priming this map with a size related to the number of finallies, e.g.? (Default is min of 7 elements, expanded by 50% when grown)

AndyAyersMS replied:
Not sure how to set this to anything more reasonable, unless you are suggesting it is oversized by default. The final size will end up being the max number of continuations for any finally in the method. Likely this value is 1 or 2 for most methods with finallys.

BruceForstall replied:
oh, that makes sense. I was thinking it would be scaled to the number of blocks, but I forgot it would get reset frequently.

BruceForstall commented:
A couple questions. Otherwise,

LGTM
